### DISCUSSION TITLE
Virtual configuration IU in p2.inf does not work in 4.0.3 (works in 3.0.5)

### DISCUSSION THREAD

**Discussion by mdaloia** [eclipse-tycho/tycho#2972] [Q&A]
Hi,

We have a "virtual" IU provided and required in a p2.inf file of a bundle to add a touchpoint configuration (it really doesn't matter what it does) that used to work in 3.0.5. We are migrating to 4.0.3 and found that it fails with:
```
[ERROR] Cannot resolve project dependencies:
[ERROR]   Software being installed: p2.inf 1.0.0.qualifier
[ERROR]   Missing requirement: p2.inf 1.0.0.qualifier requires 'org.eclipse.equinox.p2.iu; configure.p2.inf 0.0.0' but it could not be found
```

I have reduced it to an example project: https://github.com/mdaloia/tycho-p2-inf

(**Environment used**: Maven 3.9.4 / Java 17)

If I build it with
- `mvn clean package -Dtycho.localArtifacts=ignore -Dtycho-version=4.0.3`
it fails with the mentioned error and no `target/p2content.xml` is created.

I tried adding `-Dtycho.resolver.classic=false` as explained in "[mixed reactor setups require the new resolver now](https://github.com/eclipse-tycho/tycho/blob/master/RELEASE_NOTES.md#mixed-reactor-setups-require-the-new-resolver-now)"  but I got the same error.

But if I only change the version:
- `mvn clean package -Dtycho.localArtifacts=ignore -Dtycho-version=3.0.5`
it works and in the `target/p2content.xml` file I see the `<unit id='configure.p2.inf'>` and the `<required namespace='org.eclipse.equinox.p2.iu' name='configure.p2.inf' range='0.0.0'/>`

This is because I have `<pomDependencies>consider</pomDependencies>` and it shows this message:
```
[INFO] The following requirements are not satisfied yet and must be provided through pom dependencies:
[INFO]    - org.eclipse.equinox.p2.iu; configure.p2.inf 0.0.0
```
If I remove this configuration it fails with the same error as using 4.0.3.

Do I missed something or it seems to be a bug?

Thanks in advance

**Comment by laeubi** [comment]
@mdaloia this is an interesting case I assume the IU is provided by one bundle of your reactor via p2.inf?
Can you maybe try to convert your project into an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue? From technical POV I think we need to scan p2.inf for units to provide and then it should resolve the issue, I plan to publish a bugfix release of Tycho 4 in the next days so maybe this can then be included if we have a test and easy enough way to fix that.

**Reply by mdaloia** [reply]
Thanks! I will create an issue from this discussion and provide the requested integration test

> I assume the IU is provided by one bundle of your reactor via p2.inf?

yes, we have a bundle (lets call it H) that creates a virtual IU (lets call it V) just via metadata in the p2.inf of the host bundle H and adds a requirement to V in H.

We have done it in order to provide some configuration touchpoints (to modify the product .ini launcher with some arguments only if the O.S. is Linux).... but really it doesn't matter what it does, the issue is during the resolution.

I think that it can maybe be replaced with a fragment with just a p2.inf file although it will introduce more burden.

**Reply by laeubi** [reply]
p2.inf is quite tricky to handle, lets see if we can make this work, if you share some details @merks maybe knows how one can ever archive this easier.

**Reply by mdaloia** [reply]
Thanks! I've created https://github.com/eclipse-tycho/tycho/pull/2978 with the reproducer IT.
